### PR TITLE
NIFI-6576 Add HTTP authentication to confluent schema registry controller

### PR DESCRIPTION
**Pull request by teeberg**
#### Description of PR
Allow schema registry controller to connect to schema registries requiring Basic/Digest HTTP Auth.

Most of the code was written by Lakshmi Srinivas and posted as a patch under https://issues.apache.org/jira/browse/NIFI-6576, I only fixed two string comparisons.

Fixes [NIFI-6576](https://issues.apache.org/jira/browse/NIFI-6576)

I'm still having trouble running the tests locally and will keep trying.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java [hunk 1]
```diff
@@ -113,6 +115,32 @@
...
+        .required(false)
+        .sensitive(true)
+        .addValidator(StandardValidators.createRegexMatchingValidator(Pattern.compile("^[\\x20-\\x7e\\x80-\\xff]+$")))
+        .build();
+
+    static final PropertyDescriptor PROP_AUTH_TYPE = new PropertyDescriptor.Builder()
+        .name("Authentication Type")
+        .displayName("Authentication Type")
+        .description("'Basic Authentication Username' and 'Basic Authentication Password' are used if Digest Authentication is used"
+                + "for authentication.")
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:139]: There is a missing space between "usedfor" based on the String concatenation.

Comment (alopresto) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:139]: While people familiar with [RFC 2617](https://tools.ietf.org/html/rfc2617) will understand that _basic authentication_ and _digest authentication_ both use a username and password as input to an authentication process with the server, the standalone message `'Basic Authentication Username' and 'Basic Authentication Password' are used if Digest Authentication is used for authentication.` can reasonably be expected to confuse a normal user. I would recommend changing the username and password property descriptors to remove the "basic" terminology and change the description for this PD to read `Both basic authentication and digest authentication will use the 'Authentication Username' and 'Authentication Password' property values. See Confluent Schema Registry documentation for more details.` The link to said documentation is [https://docs.confluent.io/current/schema-registry/security/index.html#configuring-the-rest-api-for-basic-http-authentication](https://docs.confluent.io/current/schema-registry/security/index.html#configuring-the-rest-api-for-basic-http-authentication) but does not need to be in the property descriptor description, but can be in the processor documentation. I actually don't see any details on that page about configuring digest authentication at all -- do you have a reference for Confluent Schema Registry supporting this authentication mechanism?

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java [hunk 2]
```diff
@@ -52,6 +52,8 @@
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import java.util.regex.Pattern;
+import static org.apache.commons.lang3.StringUtils.trimToEmpty;
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:56]: I believe our coding guidelines and checkstyle discourage static imports.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java [hunk 3]
```diff
@@ -113,6 +115,32 @@
...
+        .build();
+
+    static final PropertyDescriptor PROP_AUTH_TYPE = new PropertyDescriptor.Builder()
+        .name("Authentication Type")
+        .displayName("Authentication Type")
+        .description("'Basic Authentication Username' and 'Basic Authentication Password' are used if Digest Authentication is used"
+                + "for authentication.")
+        .required(false)
+        .allowableValues("DIGEST", "BASIC")
+        .build();
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:142]: We should probably add logic in the `#customValidate()` method to ensure that if the username and password are present, the auth type property is set and valid as well. Performing the validation earlier improves the user experience and helps with debugging.

**Comment by caldiver07** [comment]
Just wanted to check up on this PR.  We are using confluent and would like to use nifi, but looks like authentication is required.  Is this PR ready to be merged.  Looks like someone has tested and appears to be working?

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java (hunk 2 above)
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:56]: I don't think static imports have ever been blocked by checkstyle. Is that something we need to verify as behavior?

**Pull request closed without merging**